lib/date_time/date_time.c

```diff
@@ -123,27 +190,137 @@ static int time_modem_get(void)
 		LOG_WRN("Did not get time from cellular network (error: %d). "
 			"This is normal as some cellular networks don't provide it or "
 			"time may not be available yet.", rc);
-		return rc;
+		return -ENODATA;
 	}
+	modem_valid_time = true;
 
 	/* Relative to 1900, as per POSIX */
 	date_time.tm_year = date_time.tm_year + 2000 - 1900;
 	/* Range is 0-11, as per POSIX */
 	date_time.tm_mon = date_time.tm_mon - 1;
 
-	if (date_time.tm_year == MODEM_TIME_DEFAULT) {
-		LOG_DBG("Modem time never set");
-		return -ENODATA;
-	}
-
 	time_aux.date_time_utc = (int64_t)timeutil_timegm64(&date_time) * 1000;
 	time_aux.last_date_time_update = k_uptime_get();
 
 	LOG_DBG("Time obtained from cellular network");
 
 	return 0;
 }
-#endif
+
+/**
+ * @brief Converts an octet having two semi-octets into a decimal.
+ *
+ * @details Semi-octet representation is explained in 3GPP TS 23.040 Section 9.1.2.3.
+ * An octet has semi-octets in the following order:
+ *   semi-octet-digit2, semi-octet-digit1
+ * Octet for decimal number '21' is hence represented as semi-octet bits:
+ *   00010010
+ *
+ * @param[in] value Octet to be converted.
+ *
+ * @return Decimal value.
+ */
+static uint8_t semioctet_to_dec(uint8_t value)
```

Comment:
Some modularity would do this library good, since its growing in size. Place these static helper/convenience functions into separate source files. Then they could be tested more easily and `date_time.c` would look a lot cleaner.

Reply:
Given the timeline, I won't have time to do this as it'll likely raise several comments and discussions for which we don't have time. We would need to leave refactoring for future.